File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/OfflineHrDataTest.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import androidx.test.espresso.matcher.ViewMatchers
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdMeasurementType
import com.polar.androidcommunications.testrules.BleLoggerTestRule
import org.hamcrest.Matchers
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.function.ThrowingRunnable
import java.io.File
import java.io.FileReader

internal class OfflineHrDataTest {

    @Rule
    @JvmField
    val bleLoggerTestRule = BleLoggerTestRule()

    @Test
    fun `test offline Type0 hr data sample`() {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     00 (raw, type 0)
        val offlineHrDataFrameHeader = byteArrayOf(
            0x0E.toByte(),
            0x00.toByte(), 0x94.toByte(), 0x35.toByte(), 0x77.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte(),
            0x00.toByte(),
        )
        val previousTimeStamp = 0uL

        // index                                                   data:
        // 0             sample0                                   00
        val expectedSample0 = 0
        // 1             sample0                                   FF
        val expectedSample1 = 255
        // last index    sampleN                                   7F
        val expectedSampleLast = 127
        val expectedSampleSize = 9
        val offlineHrDataFrameContent = byteArrayOf(
            0x00.toByte(), 0xFF.toByte(), 0x32.toByte(), 0x32.toByte(), 0x33.toByte(), 0x33.toByte(), 0x34.toByte(), 0x35.toByte(), 0x7F.toByte(),
        )

        val dataFrame = PmdDataFrame(
            data = offlineHrDataFrameHeader + offlineHrDataFrameContent,
            getPreviousTimeStamp = { pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> previousTimeStamp },
            getFactor = { 1.0f }
        ) { 0 }


        // Act
        val offlineHrData = OfflineHrData.parseDataFromDataFrame(dataFrame)

        // Assert
        Assert.assertEquals(expectedSampleSize, offlineHrData.hrSamples.size)
        Assert.assertEquals(expectedSample0, offlineHrData.hrSamples.first().hr)
        Assert.assertEquals(expectedSample1, offlineHrData.hrSamples[1].hr)
        Assert.assertEquals(expectedSampleLast, offlineHrData.hrSamples.last().hr)
    }

    @Test
    fun `test offline Type0 compressed hr data sample throws`() {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     00 (raw, type 0)
        val offlineHrDataFrameHeader = byteArrayOf(
            0x0E.toByte(),
            0x00.toByte(), 0x94.toByte(), 0x35.toByte(), 0x77.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte(),
            0x80.toByte(),
        )
        val previousTimeStamp = 0uL

        val dataFrame = PmdDataFrame(
            data = offlineHrDataFrameHeader,
            getPreviousTimeStamp = { pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> previousTimeStamp },
            getFactor = { 1.0f }
        ) { 0 }

        var throwingRunnable = ThrowingRunnable { OfflineHrData.parseDataFromDataFrame(dataFrame) }
        val exception = Assert.assertThrows(Exception::class.java, throwingRunnable)
        ViewMatchers.assertThat(
            exception.message,
            Matchers.equalTo("Compressed FrameType: TYPE_0 is not supported by Offline HR data parser")
        )
    }

    @Test
    fun `test offline type 1 hr data sample`() {

        val offlineHrDataFrameHeader = byteArrayOf(
            0x14.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(),
            0x0.toByte(), 0x0.toByte(), 0x1.toByte()
        )

        val offlineHrDataFrameContent = byteArrayOf(
            72.toByte(), 86.toByte(), 71.toByte(), 81.toByte(), 64.toByte(), 82.toByte()
        )

        val dataFrame = PmdDataFrame(
            data = offlineHrDataFrameHeader + offlineHrDataFrameContent,
            getPreviousTimeStamp = {  pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> 0uL },
            getFactor = { 1.0f }
        ) { 0 }

        val offlineHrData = OfflineHrData.parseDataFromDataFrame(dataFrame)

        Assert.assertEquals(2, offlineHrData.hrSamples.size)
        Assert.assertEquals(72, offlineHrData.hrSamples.first().hr)
        Assert.assertEquals(86, offlineHrData.hrSamples.first().ppgQuality)
        Assert.assertEquals(71, offlineHrData.hrSamples.first().correctedHr)
        Assert.assertEquals(81, offlineHrData.hrSamples.last().hr)
        Assert.assertEquals(64, offlineHrData.hrSamples.last().ppgQuality)
        Assert.assertEquals(82, offlineHrData.hrSamples.last().correctedHr)
    }

    @Test
    fun `test offline Type1 compressed hr data sample throws`() {

        val offlineHrDataFrameHeader = byteArrayOf(
            0x14.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(), 0x0.toByte(),
            0x0.toByte(), 0x0.toByte(), 0x81.toByte()
        )
        val previousTimeStamp = 0uL

        val dataFrame = PmdDataFrame(
            data = offlineHrDataFrameHeader,
            getPreviousTimeStamp = { pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> previousTimeStamp },
            getFactor = { 1.0f }
        ) { 0 }

        var throwingRunnable = ThrowingRunnable { OfflineHrData.parseDataFromDataFrame(dataFrame) }
        val exception = Assert.assertThrows(Exception::class.java, throwingRunnable)
        ViewMatchers.assertThat(
            exception.message,
            Matchers.equalTo("Compressed FrameType: TYPE_1 is not supported by Offline HR data parser")
        )
    }

    @Test
    fun offlineHrGoldenVectors_matchAndroidBehavior() {
        val vectors = loadOfflineHrVectors()
        Assert.assertTrue("Expected offline HR golden vectors", vectors.isNotEmpty())

        vectors.forEach { vector ->
            val caseId = vector.get("id").asString
            if (vector.getAsJsonObject("platforms")?.get("android")?.asBoolean == false) {
                return@forEach
            }
            val input = vector.getAsJsonObject("input")
            val expected = vector.getAsJsonObject("expected")
            val frame = PmdDataFrame(
                data = input.get("dataFrameHex").asString.hexToByteArray(),
                getPreviousTimeStamp = { _, _ -> input.get("previousTimeStamp").asLong.toULong() },
                getFactor = { input.get("factor").asFloat },
                getSampleRate = { input.get("sampleRate").asInt }
            )

            if (expected.has("parseError")) {
                assertParseError(caseId, expected.get("parseError").asString, frame)
                return@forEach
            }

            val offlineHrData = OfflineHrData.parseDataFromDataFrame(frame)

            assertOfflineHrSamples(caseId, expected.getAsJsonArray("samples"), offlineHrData.hrSamples)
        }
    }

    private fun assertParseError(caseId: String, expectedError: String, frame: PmdDataFrame) {
        when (expectedError) {
            "unsupportedFrame" -> Assert.assertThrows(caseId, Exception::class.java) {
                OfflineHrData.parseDataFromDataFrame(frame)
            }
            "unsupportedCompressedFrame" -> Assert.assertThrows(caseId, Exception::class.java) {
                OfflineHrData.parseDataFromDataFrame(frame)
            }
            "malformedFrame" -> Assert.assertThrows(caseId, Exception::class.java) {
                OfflineHrData.parseDataFromDataFrame(frame)
            }
            else -> Assert.fail("$caseId has unsupported parse error expectation $expectedError")
        }
    }

    @Test
    fun `offline HR golden vectors follow neutral KMP vector shape`() {
        loadOfflineHrVectors().forEach { vector ->
            val id = vector.get("id").asString
            Assert.assertTrue(id, vector.has("area"))
            Assert.assertTrue(id, vector.has("case"))
            Assert.assertTrue(id, vector.has("source"))
            Assert.assertTrue(id, vector.has("input"))
            Assert.assertTrue(id, vector.has("expected"))
            Assert.assertTrue(id, vector.has("platforms"))
            val platforms = vector.getAsJsonObject("platforms")
            Assert.assertTrue(id, platforms.has("android"))
            Assert.assertTrue(id, platforms.has("ios"))
            Assert.assertTrue(id, platforms.has("common"))
        }
    }

    @Test
    fun `offline HR readiness manifest is pinned before parser migration`() {
        val manifest = loadOfflineHrReadinessManifest()
        val id = manifest.get("id").asString
        val input = manifest.getAsJsonObject("input")
        val expected = manifest.getAsJsonObject("expected")
        val consumerTests = manifest.getAsJsonObject("consumerTests")
        val requiredFamilies = input.getAsJsonArray("requiredBehaviorFamilies").map { it.asString }
        val coveredFamilies = expected.getAsJsonArray("coveredBehaviorFamilies").map { it.asString }
        val policyVectorPaths = input.getAsJsonArray("policyVectorPaths").map { it.asString }

        Assert.assertEquals("offline-hr-readiness", id)
        Assert.assertEquals("offlineHrReadiness", input.get("kind").asString)
        Assert.assertEquals(id, OFFLINE_HR_READINESS_POLICY_VECTOR_PATHS, policyVectorPaths)
        Assert.assertEquals(id, OFFLINE_HR_READINESS_FAMILIES, requiredFamilies)
        Assert.assertEquals(id, OFFLINE_HR_READINESS_FAMILIES, coveredFamilies)
        Assert.assertEquals(id, listOf("com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model.OfflineHrDataTest"), consumerTests.getAsJsonArray("android").map { it.asString })
        Assert.assertEquals(id, listOf("OfflineHrDataTest"), consumerTests.getAsJsonArray("ios").map { it.asString })
        Assert.assertEquals(id, listOf("com.polar.sharedtest.OfflineHrParserCommonPolicyTest"), consumerTests.getAsJsonArray("commonPrototype").map { it.asString })
    }

    private fun assertOfflineHrSamples(
        caseId: String,
        expectedSamples: JsonArray,
        actualSamples: List<OfflineHrData.OfflineHrSample>
    ) {
        Assert.assertEquals(caseId, expectedSamples.size(), actualSamples.size)
        expectedSamples.forEachIndexed { index, expectedSample ->
            val sample = expectedSample.asJsonObject
            val actualSample = actualSamples[index]
            Assert.assertEquals(caseId, sample.get("hr").asInt, actualSample.hr)
            Assert.assertEquals(caseId, sample.get("ppgQuality").asInt, actualSample.ppgQuality)
            Assert.assertEquals(caseId, sample.get("correctedHr").asInt, actualSample.correctedHr)
        }
    }

    private fun loadOfflineHrVectors(): List<JsonObject> {
        val vectorDirectory = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/sensors")
        return vectorDirectory
            .listFiles { file -> file.isFile && file.extension == "json" && file.name.startsWith("offline-hr-") }
            .orEmpty()
            .sortedBy { it.name }
            .map { file ->
                FileReader(file).use { reader ->
                    JsonParser().parse(reader).asJsonObject
                }
            }
            .filter { vector -> vector.getAsJsonObject("input").get("kind")?.asString != "offlineHrReadiness" }
    }

    private fun loadOfflineHrReadinessManifest(): JsonObject {
        val vectorFile = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/sensors/offline-hr-readiness.json")
        FileReader(vectorFile).use { reader ->
            return JsonParser().parse(reader).asJsonObject
        }
    }

    private fun findRepositoryRoot(): File {
        val userDirectory = System.getProperty("user.dir") ?: error("user.dir is not set")
        var directory = File(userDirectory).absoluteFile
        while (true) {
            if (directory.resolve("testdata/golden-vectors").isDirectory) {
                return directory
            }
            directory = directory.parentFile ?: error("Could not find repository root from $userDirectory")
        }
    }

    private fun String.hexToByteArray(): ByteArray {
        require(length % 2 == 0) { "Hex string must have an even length" }
        return chunked(2)
            .map { it.toInt(16).toByte() }
            .toByteArray()
    }

    private companion object {
        val OFFLINE_HR_READINESS_POLICY_VECTOR_PATHS = listOf(
            "protocol/sensors/offline-hr-compressed-type0-unsupported.json",
            "protocol/sensors/offline-hr-raw-type0-empty.json",
            "protocol/sensors/offline-hr-raw-type0-hr-only-boundaries.json",
            "protocol/sensors/offline-hr-raw-type1-truncated-tuple-android-error.json",
            "protocol/sensors/offline-hr-raw-type1-two-samples.json",
            "protocol/sensors/offline-hr-raw-type2-unsupported.json"
        )

        val OFFLINE_HR_READINESS_FAMILIES = listOf(
            "raw-type0-hr-only-samples",
            "raw-type0-empty-recording",
            "raw-type1-hr-ppg-quality-corrected-hr-triples",
            "unsupported-compressed-frame-policy",
            "unsupported-raw-frame-policy",
            "truncated-raw-type1-tuple-policy",
            "platform-offline-hr-vector-reference-gate",
            "compile-verification-gate"
        )
    }
}

File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/OfflineHrDataTest.swift
//  Copyright © 2023 Polar. All rights reserved.

import XCTest
#if canImport(PolarBleSdkShared)
import PolarBleSdkShared
#endif
@testable import iOSCommunications

final class OfflineHrDataTest: XCTestCase {
    
    func testUncompressedOfflineHrDataFrameType0() throws {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     00 (raw, type 0)
        let offlineHrDataFrameHeader = Data([
            0x0E,
            0x00, 0x94, 0x35, 0x77, 0x00, 0x00, 0x00, 0x00,
            0x00,
        ])
        let previousTimeStamp: UInt64 = 0
        
        // index                                                   data:
        // 0             sample0                                   00
        let expectedSample0: UInt8 = 0
        // 1             sample0                                   FF
        let expectedSample1: UInt8 = 255
        // last index    sampleN                                   7F
        let expectedSampleLast: UInt8 = 127
        let expectedSampleSize = 9
        let offlineHrDataFrameContent = Data([
            0x00, 0xFF, 0x32, 0x32, 0x33, 0x33, 0x34, 0x35, 0x7F,
        ])
        
        let factor:Float = 1.0
        let dataFrame = try PmdDataFrame(
            data: offlineHrDataFrameHeader + offlineHrDataFrameContent,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })
        
        
        // Act
        let offlineHrData = try OfflineHrData.parseDataFromDataFrame(frame: dataFrame)
        
        // Assert
        XCTAssertEqual(expectedSampleSize, offlineHrData.samples.count)
        XCTAssertEqual(expectedSample0, offlineHrData.samples.first?.hr)
        XCTAssertEqual(expectedSample1, offlineHrData.samples[1].hr)
        XCTAssertEqual(expectedSampleLast, offlineHrData.samples.last?.hr)
    }

    func testCompressedOfflineHrDataFrameType0ThrowsError() throws {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     80 (compressed, type 0)
        let offlineHrDataFrameHeader = Data([
            0x0E, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
            0x0, 0x0, 0x80,
        ])
        let previousTimeStamp: UInt64 = 0
        let factor:Float = 1.0
        
        // Act
        let dataFrame = try PmdDataFrame(
            data: offlineHrDataFrameHeader,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })

        // Assert
        XCTAssertThrowsError(try OfflineHrData.parseDataFromDataFrame(frame: dataFrame), "Raw FrameType: TYPE_0 is not supported by PPG data parser")
        XCTAssertThrowsError(try OfflineHrData.parseDataFromDataFrame(frame: dataFrame)) { error in
            guard case BleGattException.gattDataError = error else {
                return XCTFail()
            }
        }
    }

    func testUncompressedOfflineHrDataFrameType1() throws {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     01 (raw, type 1)
        let offlineHrDataFrameHeader = Data([
            0x0E, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
            0x0, 0x0, 0x1,
        ])
        let previousTimeStamp: UInt64 = 0

        // index                                                   data:
        // 0             sample0                                   48
        let expectedHR1: UInt8 = 72
        // 3             sample0                                   51
        let expectedHR2: UInt8 = 81
        // 1             sample0                                   56
        let expectedPPGQuality1: UInt8 = 86
        // 4             sample0                                   40
        let expectedPPGQuality2: UInt8 = 64
        // 2             sample0                                   47
        let expectedCorrectedHR1: UInt8 = 71
        // 5             sample0                                   52
        let expectedCorrectedHR2: UInt8 = 82
        let expectedSampleSize = 2
        
        let offlineHrDataFrameContent = Data([
            0x48, 0x56, 0x47, 0x51, 0x40, 0x52,
        ])

        let factor:Float = 1.0
        let dataFrame = try PmdDataFrame(
            data: offlineHrDataFrameHeader + offlineHrDataFrameContent,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })

        // Act
        let offlineHrData = try OfflineHrData.parseDataFromDataFrame(frame: dataFrame)

        // Assert
        XCTAssertEqual(expectedSampleSize, offlineHrData.samples.count)
        XCTAssertEqual(expectedHR1, offlineHrData.samples.first?.hr)
        XCTAssertEqual(expectedHR2, offlineHrData.samples.last?.hr)
        XCTAssertEqual(expectedPPGQuality1, offlineHrData.samples.first?.ppgQuality)
        XCTAssertEqual(expectedPPGQuality2, offlineHrData.samples.last?.ppgQuality)
        XCTAssertEqual(expectedCorrectedHR1, offlineHrData.samples.first?.correctedHr)
        XCTAssertEqual(expectedCorrectedHR2, offlineHrData.samples.last?.correctedHr)
    }

    func testOfflineHrRawParserUsesSharedKmpWhenLinked() throws {
        #if canImport(PolarBleSdkShared)
        let dataFrameHex = "0e000000000000000001485647514052"
        let sharedRows = try XCTUnwrap(OfflineHrDataRuntimePlanner.rawSamples(dataFrameHex: dataFrameHex, previousTimeStamp: 0, factor: 1.0, sampleRate: 0))
        XCTAssertFalse(sharedRows.isEmpty)

        let dataFrame = try PmdDataFrame(
            data: Data(hexString: dataFrameHex),
            { _, _ in 0 },
            { _ in 1.0 },
            { _ in 0 })
        let offlineHrData = try OfflineHrData.parseDataFromDataFrame(frame: dataFrame)
        let sharedSamples = try sharedRows.split(separator: "|").map { row -> (UInt8, UInt8, UInt8) in
            let fields = row.split(separator: ",")
            return (
                try XCTUnwrap(UInt8(fields[0])),
                try XCTUnwrap(UInt8(fields[1])),
                try XCTUnwrap(UInt8(fields[2]))
            )
        }

        XCTAssertEqual(sharedSamples.count, offlineHrData.samples.count)
        for (index, sharedSample) in sharedSamples.enumerated() {
            XCTAssertEqual(sharedSample.0, offlineHrData.samples[index].hr)
            XCTAssertEqual(sharedSample.1, offlineHrData.samples[index].ppgQuality)
            XCTAssertEqual(sharedSample.2, offlineHrData.samples[index].correctedHr)
        }
        #else
        throw XCTSkip("PolarBleSdkShared is not linked in this build")
        #endif
    }

    func testCompressedOfflineHrDataFrameType1ThrowsError() throws {
        // Arrange
        // HEX: 0E 00 00 00 00 00 00 00 00 00
        // index                                                   data:
        // 0        type                                           0E (Offline hr)
        // 1..9     timestamp                                      00 00 00 00 00 00 00 00
        // 10       frame type                                     81 (compressed, type 1)
        let offlineHrDataFrameHeader = Data([
            0x0E, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
            0x0, 0x0, 0x81,
        ])
        let previousTimeStamp: UInt64 = 0

        let factor:Float = 1.0
        
        // Act
        let dataFrame = try PmdDataFrame(
            data: offlineHrDataFrameHeader,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })

        // Assert
        XCTAssertThrowsError(try PpgData.parseDataFromDataFrame(frame: dataFrame), "Raw FrameType: TYPE_1 is not supported by PPG data parser")
        XCTAssertThrowsError(try PpgData.parseDataFromDataFrame(frame: dataFrame)) { error in
            guard case BleGattException.gattDataError = error else {
                return XCTFail()
            }
        }
    }

    func testOfflineHrGoldenVectorsMatchIOSCommunicationsBehavior() throws {
        let vectors = try loadOfflineHrGoldenVectors()
        XCTAssertFalse(vectors.isEmpty, "Expected offline HR golden vectors")

        for vector in vectors {
            let id = vector["id"] as? String ?? "unknown-vector"
            if let platforms = vector["platforms"] as? [String: Any],
               let supported = platforms["ios"] as? Bool,
               !supported {
                continue
            }
            let input = try XCTUnwrap(vector["input"] as? [String: Any], id)
            let expected = try XCTUnwrap(vector["expected"] as? [String: Any], id)
            let dataFrame = try PmdDataFrame(
                data: Data(hexString: try XCTUnwrap(input["dataFrameHex"] as? String, id)),
                { _, _ in UInt64(truncating: input["previousTimeStamp"] as? NSNumber ?? 0) },
                { _ in Float(truncating: input["factor"] as? NSNumber ?? 1.0) },
                { _ in UInt(truncating: input["sampleRate"] as? NSNumber ?? 0) })

            if let parseError = expected["parseError"] as? String {
                XCTAssertThrowsError(try OfflineHrData.parseDataFromDataFrame(frame: dataFrame), id) { error in
                    switch parseError {
                    case "unsupportedFrame":
                        guard case BleGattException.gattDataError = error else {
                            return XCTFail("Expected gattDataError for \(id), got \(error)")
                        }
                    case "unsupportedCompressedFrame":
                        guard case BleGattException.gattDataError = error else {
                            return XCTFail("Expected gattDataError for \(id), got \(error)")
                        }
                    case "malformedFrame":
                        guard case BleGattException.gattDataError = error else {
                            return XCTFail("Expected gattDataError for \(id), got \(error)")
                        }
                    default:
                        XCTFail("Unsupported parse error expectation in \(id): \(parseError)")
                    }
                }
                continue
            }

            let offlineHrData = try OfflineHrData.parseDataFromDataFrame(frame: dataFrame)

            let samples = try XCTUnwrap(expected["samples"] as? [[String: Any]], id)
            XCTAssertEqual(samples.count, offlineHrData.samples.count, id)
            for (index, expectedSample) in samples.enumerated() {
                let actualSample = offlineHrData.samples[index]
                XCTAssertEqual(try XCTUnwrap(expectedSample["hr"] as? NSNumber, id).uint8Value, actualSample.hr, id)
                XCTAssertEqual(try XCTUnwrap(expectedSample["ppgQuality"] as? NSNumber, id).uint8Value, actualSample.ppgQuality, id)
                XCTAssertEqual(try XCTUnwrap(expectedSample["correctedHr"] as? NSNumber, id).uint8Value, actualSample.correctedHr, id)
            }
        }
    }

    func testOfflineHrGoldenVectorsFollowNeutralKmpShape() throws {
        for vector in try loadOfflineHrGoldenVectors() {
            let id = try XCTUnwrap(vector["id"] as? String)
            XCTAssertNotNil(vector["area"], id)
            XCTAssertNotNil(vector["case"], id)
            XCTAssertNotNil(vector["source"], id)
            XCTAssertNotNil(vector["input"], id)
            XCTAssertNotNil(vector["expected"], id)
            let platforms = try XCTUnwrap(vector["platforms"] as? [String: Any], id)
            XCTAssertNotNil(platforms["android"], id)
            XCTAssertNotNil(platforms["ios"], id)
            XCTAssertNotNil(platforms["common"], id)
        }
    }

    func testOfflineHrReadinessManifestIsPinnedBeforeParserMigration() throws {
        let manifest = try loadOfflineHrReadinessManifest()
        let id = try XCTUnwrap(manifest["id"] as? String)
        let input = try XCTUnwrap(manifest["input"] as? [String: Any], id)
        let expected = try XCTUnwrap(manifest["expected"] as? [String: Any], id)
        let consumerTests = try XCTUnwrap(manifest["consumerTests"] as? [String: Any], id)
        let requiredFamilies = try XCTUnwrap(input["requiredBehaviorFamilies"] as? [String], id)
        let coveredFamilies = try XCTUnwrap(expected["coveredBehaviorFamilies"] as? [String], id)
        let policyVectorPaths = try XCTUnwrap(input["policyVectorPaths"] as? [String], id)

        XCTAssertEqual("offline-hr-readiness", id)
        XCTAssertEqual("offlineHrReadiness", input["kind"] as? String, id)
        XCTAssertEqual(OFFLINE_HR_READINESS_POLICY_VECTOR_PATHS, policyVectorPaths, id)
        XCTAssertEqual(OFFLINE_HR_READINESS_FAMILIES, requiredFamilies, id)
        XCTAssertEqual(OFFLINE_HR_READINESS_FAMILIES, coveredFamilies, id)
        XCTAssertEqual(["com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model.OfflineHrDataTest"], consumerTests["android"] as? [String], id)
        XCTAssertEqual(["OfflineHrDataTest"], consumerTests["ios"] as? [String], id)
        XCTAssertEqual(["com.polar.sharedtest.OfflineHrParserCommonPolicyTest"], consumerTests["commonPrototype"] as? [String], id)
    }

    private func loadOfflineHrGoldenVectors() throws -> [[String: Any]] {
        let vectorDirectory = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/sensors")
        return try FileManager.default
            .contentsOfDirectory(at: vectorDirectory, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" && $0.lastPathComponent.hasPrefix("offline-hr-") }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
            .map { file in
                let data = try Data(contentsOf: file)
                return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], file.path)
            }
            .filter { vector in
                guard let input = vector["input"] as? [String: Any] else {
                    return true
                }
                return input["kind"] as? String != "offlineHrReadiness"
            }
    }

    private func loadOfflineHrReadinessManifest() throws -> [String: Any] {
        let vectorFile = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/sensors/offline-hr-readiness.json")
        let data = try Data(contentsOf: vectorFile)
        return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], vectorFile.path)
    }


    private let OFFLINE_HR_READINESS_POLICY_VECTOR_PATHS = [
        "protocol/sensors/offline-hr-compressed-type0-unsupported.json",
        "protocol/sensors/offline-hr-raw-type0-empty.json",
        "protocol/sensors/offline-hr-raw-type0-hr-only-boundaries.json",
        "protocol/sensors/offline-hr-raw-type1-truncated-tuple-android-error.json",
        "protocol/sensors/offline-hr-raw-type1-two-samples.json",
        "protocol/sensors/offline-hr-raw-type2-unsupported.json"
    ]

    private let OFFLINE_HR_READINESS_FAMILIES = [
        "raw-type0-hr-only-samples",
        "raw-type0-empty-recording",
        "raw-type1-hr-ppg-quality-corrected-hr-triples",
        "unsupported-compressed-frame-policy",
        "unsupported-raw-frame-policy",
        "truncated-raw-type1-tuple-policy",
        "platform-offline-hr-vector-reference-gate",
        "compile-verification-gate"
    ]
}

private extension Data {
    init(hexString: String) throws {
        guard hexString.count.isMultiple(of: 2) else {
            throw NSError(domain: "OfflineHrDataTest", code: 2, userInfo: [NSLocalizedDescriptionKey: "Hex string must have an even length"])
        }
        var bytes: [UInt8] = []
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let nextIndex = hexString.index(index, offsetBy: 2)
            let byteString = String(hexString[index..<nextIndex])
            guard let byte = UInt8(byteString, radix: 16) else {
                throw NSError(domain: "OfflineHrDataTest", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid hex byte \(byteString)"])
            }
            bytes.append(byte)
            index = nextIndex
        }
        self.init(bytes)
    }
}
